crates/sc-subspace-block-relay/src/consensus/relay.rs

```diff
@@ -496,6 +548,27 @@ where
         })
     }
 
+    /// Handles version mismatch detected when the initial request
+    /// was received.
+    fn handle_version_mismatch(
+        &self,
+        block_hash: &BlockHash<Block>,
+        block_attributes: BlockAttributes,
+    ) -> Result<Vec<u8>, RelayError> {
+        // Return full block on version mismatch.
+        let partial_block = self.get_partial_block(block_hash, block_attributes)?;
+        let body = if block_attributes.contains(BlockAttributes::BODY) {
+            Some(block_transactions(block_hash, self.client.as_ref())?)
+        } else {
+            None
+        };
+        let initial_response: InitialResponse<Block, ProtoServer::Response> = InitialResponse {
+            block_hash: *block_hash,
+            response: BlockResponse::Complete(partial_block.block_data(body)),
```

Comment:
IMO, this is quite unexpected. I dont think it make sense to send full block when the client specifically asked for a protocol response. If the server doe not support, I would error out and let the client decide how they want to proceed.

They can choose to go with full block but server should not make that decision for them TBH.

Reply:
Goal is to avoid erroring out as much as possible, which I feel would create an unstable/hard to debug scenario during upgrade - substrate would need to punt to next peer, etc. Keep in mind this happens only during upgrade(transient) period, returning the block keeps is more deterministic.

Reply:
>Goal is to avoid erroring out as much as possible, which I feel would create an unstable/hard to debug scenario during upgrade - substrate would need to punt to next peer, etc. Keep in mind this happens only during upgrade(transient) period, returning the block keeps it more deterministic.

I think avoid the erroring is hard to debug TBH. Returning a full block when the protocol mismatched IMHO is not correct. Ideally, full block could be moved to its own protocol and client can pick and chose which protocol to use. If the protocol is mismatched, then they will move to full download protocol instead. In the current case, seems like a protocol is just optional since I would get a full block irrespective

Reply:
Where do you see an extra roundtrip?

Reply:
> If the protocol is mismatched, then they will move to full download protocol instead

Hm, that was my understanding here. How does the client know the server protocol/version otherwise?

Reply:
> Hm, that was my understanding here. How does the client know the server protocol/version otherwise?

That is a possibility, we don't have to do it that way and I would say we shouldn't. We just do staged upgrades by first introducing support for new protocol and then starting using it later.